internal/controlplane/handlers_entity_instances.go

```diff
@@ -0,0 +1,311 @@
+// SPDX-FileCopyrightText: Copyright 2024 The Minder Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package controlplane
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/mindersec/minder/internal/db"
+	"github.com/mindersec/minder/internal/engine/engcontext"
+	"github.com/mindersec/minder/internal/engine/entities"
+	"github.com/mindersec/minder/internal/logger"
+	"github.com/mindersec/minder/internal/util"
+	pb "github.com/mindersec/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// ListEntities returns a list of entity instances for a given project and provider
+func (s *Server) ListEntities(
+	ctx context.Context,
+	in *pb.ListEntitiesRequest,
+) (*pb.ListEntitiesResponse, error) {
+	entityCtx := engcontext.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+	providerName := entityCtx.Provider.Name
+
+	logger.BusinessRecord(ctx).Provider = providerName
+	logger.BusinessRecord(ctx).Project = projectID
+
+	// Get provider ID from name
+	provider, err := s.providerStore.GetByName(ctx, projectID, providerName)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "provider not found")
+		}
+		return nil, fmt.Errorf("error getting provider: %w", err)
+	}
+
+	// Get entity type from request
+	entityType := in.GetEntityType()
+	if entityType == pb.Entity_ENTITY_UNSPECIFIED {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "entity type must be specified")
+	}
+
+	// Get limit from request
+	limit := in.GetCursor().GetSize()
+	if limit == 0 {
+		limit = 20 // Default limit
+	}
+
+	// Get cursor from request
+	cursor := in.GetCursor().GetCursor()
+
+	// Call service to get entities
+	outentities, nextCursor, err := s.entityService.ListEntities(
+		ctx,
+		projectID,
+		provider.ID,
+		entityType,
+		cursor,
+		int64(limit),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create response
+	resp := &pb.ListEntitiesResponse{
+		Results: outentities,
+		Page: &pb.CursorPage{
+			Next: &pb.Cursor{
+				Cursor: nextCursor,
+				Size:   limit,
+			},
+		},
+	}
+
+	return resp, nil
+}
+
+// GetEntityById returns an entity instance for a given entity ID
+func (s *Server) GetEntityById(
+	ctx context.Context,
+	in *pb.GetEntityByIdRequest,
+) (*pb.GetEntityByIdResponse, error) {
+	// Parse entity ID
+	entityID, err := uuid.Parse(in.GetId())
+	if err != nil {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "invalid entity ID")
+	}
+
+	projectID := GetProjectID(ctx)
```

Comment:
Interesting that List and Get-by-Name require a provider and a type, but Get-by-ID does not.  Might be worth documenting that.

Reply:
I actually still don't know how to feel about that. In a multi-provider world, getting by ID might be an easy way to fetch an entity for a particular provider. Perhaps I should allow for further filtering with provider ID.